"buildkite-agent upload" crashes with SIGSEGV: segmentation violation

Just noticed this on our Mac machines where I upgrade to Buildkite Agent 3.8.0:

```
buildkite-agent artifact upload src/test/shell/bazel/jdeps_test/test.log
--
  | 2019-01-10 14:58:29 INFO   Found 1 files that match "src/test/shell/bazel/jdeps_test/test.log"
  | panic: runtime error: invalid memory address or nil pointer dereference
  | [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12e0572]
  |  
  | goroutine 1 [running]:
  | github.com/buildkite/agent/logger.(*Logger).Info(0x0, 0x1ad0c44, 0x1d, 0xc00021cb60, 0x3, 0x3)
  | /code/logger/log.go:94 +0x22
  | github.com/buildkite/agent/agent.(*ArtifactBatchCreator).Create(0xc0000d0ff0, 0xc0000d0ff0, 0xc000280000, 0x86, 0x0, 0x0)
  | /code/agent/artifact_batch_creator.go:49 +0x30b
  | github.com/buildkite/agent/agent.(*ArtifactUploader).upload(0xc00001d840, 0xc00000c1f0, 0x1, 0x1, 0x2, 0x2)
  | /code/agent/artifact_uploader.go:211 +0x32b
  | github.com/buildkite/agent/agent.(*ArtifactUploader).Upload(0xc00001d840, 0xc00002e059, 0x1e)
  | /code/agent/artifact_uploader.go:55 +0x1ec
  | github.com/buildkite/agent/clicommand.glob..func5(0xc00010cdc0)
  | /code/clicommand/artifact_upload.go:95 +0x2dc
  | github.com/urfave/cli.HandleAction(0x18ba680, 0x1af3ff8, 0xc00010cdc0, 0x0, 0xc0000a73e0)
  | /go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/app.go:503 +0x7c
  | github.com/urfave/cli.Command.Run(0x1aba774, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1ad6f77, 0x23, 0x0, ...)
  | /go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/command.go:165 +0x459
  | github.com/urfave/cli.(*App).RunAsSubcommand(0xc0001f2380, 0xc00010cb00, 0x0, 0x0)
  | /go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/app.go:383 +0x827
  | github.com/urfave/cli.Command.startApp(0x1abc02d, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1adee80, 0x2d, 0x0, ...)
  | /go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/command.go:330 +0x808
  | github.com/urfave/cli.Command.Run(0x1abc02d, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1adee80, 0x2d, 0x0, ...)
  | /go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/command.go:103 +0x80f
  | github.com/urfave/cli.(*App).Run(0xc0001f21c0, 0xc00001c100, 0x4, 0x4, 0x0, 0x0)
  | /go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/app.go:259 +0x6bb
  | main.main()
  | /code/main.go:109 +0x4ff
  | Traceback (most recent call last):
  | File "bazelci.py", line 1762, in <module>
  | sys.exit(main())
  | File "bazelci.py", line 1746, in main
  | incompatible_flags=args.incompatible_flag,
  | File "bazelci.py", line 551, in execute_commands
  | upload_test_logs(test_bep_file, tmpdir)
  | File "bazelci.py", line 968, in upload_test_logs
  | execute_command(["buildkite-agent", "artifact", "upload", ";".join(test_logs)])
  | File "bazelci.py", line 1034, in execute_command
  | return subprocess.run(args, shell=shell, check=fail_if_nonzero, env=os.environ).returncode
  | File "/usr/local/Cellar/python/3.7.2/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 487, in run
  | output=stdout, stderr=stderr)
  | subprocess.CalledProcessError: Command '['buildkite-agent', 'artifact', 'upload', 'src/test/shell/bazel/jdeps_test/test.log']' returned non-zero exit status 2.
  | 🚨 Error: The command exited with status 1

```

Does that stack trace tell you enough, or should I do more debugging? :)

Same problem here, also with artifact upload:
```
$ buildkite-agent artifact upload build.tar.gz
2019-01-10 15:20:17 INFO   Found 1 files that match "build.tar.gz"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6d8db2]
 
goroutine 1 [running]:
github.com/buildkite/agent/logger.(*Logger).Info(0x0, 0xed0a77, 0x1d, 0xc0001fab60, 0x3, 0x3)
	/code/logger/log.go:94 +0x22
github.com/buildkite/agent/agent.(*ArtifactBatchCreator).Create(0xc0000a6ff0, 0xc0000a6ff0, 0x0, 0x0, 0x0, 0x0)
	/code/agent/artifact_batch_creator.go:49 +0x30b
github.com/buildkite/agent/agent.(*ArtifactUploader).upload(0xc00001d880, 0xc00000c200, 0x1, 0x1, 0x2, 0x2)
	/code/agent/artifact_uploader.go:211 +0x32b
github.com/buildkite/agent/agent.(*ArtifactUploader).Upload(0xc00001d880, 0xc000026019, 0x1e)
	/code/agent/artifact_uploader.go:55 +0x1ec
github.com/buildkite/agent/clicommand.glob..func5(0xc0000ed080)
	/code/clicommand/artifact_upload.go:95 +0x2dc
github.com/urfave/cli.HandleAction(0xcba6e0, 0xef3cf8, 0xc0000ed080, 0x0, 0xc000081440)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/app.go:503 +0x7c
github.com/urfave/cli.Command.Run(0xeba148, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xed6d3f, 0x23, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/command.go:165 +0x459
github.com/urfave/cli.(*App).RunAsSubcommand(0xc0001d2380, 0xc0000ecdc0, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/app.go:383 +0x827
github.com/urfave/cli.Command.startApp(0xebba16, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0xedebcf, 0x2d, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/command.go:330 +0x808
github.com/urfave/cli.Command.Run(0xebba16, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0xedebcf, 0x2d, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/command.go:103 +0x80f
github.com/urfave/cli.(*App).Run(0xc0001d21c0, 0xc00001c080, 0x4, 0x4, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20180226030253-8e01ec4cd3e2/app.go:259 +0x6bb
main.main()
	/code/main.go:109 +0x4ff
🚨 Error: Error running `/usr/bin/buildkite-agent artifact upload build.tar.gz`: exit status 2
```

We are also rolling back to 3.7.0 to make sure it wasn't something else, but we didn't have any problems until we picked up the new version.

--- a/agent/artifact_uploader.go
+++ b/agent/artifact_uploader.go
@@ -207,6 +207,7 @@ func (a *ArtifactUploader) upload(artifacts []*api.Artifact) error {
 		JobID:             a.JobID,
 		Artifacts:         artifacts,
 		UploadDestination: a.Destination,
+		Logger:            a.Logger,
 	}
 	artifacts, err = batchCreator.Create()
 	if err != nil {

--- a/clicommand/agent_start.go
+++ b/clicommand/agent_start.go
@@ -452,6 +452,7 @@ var AgentStartCommand = cli.Command{
 			MetricsCollector: &metrics.Collector{
 				Datadog:     cfg.MetricsDatadog,
 				DatadogHost: cfg.MetricsDatadogHost,
+				Logger:      l,
 			},
 		}
 

